### DISCUSSION TITLE
How to extend Saleor e-commerce to an entire new software, where the e-commerce will be only part of it?

### DISCUSSION THREAD

**Discussion by leandrobortolotto** [saleor/saleor#12727] [Q&A] [answered]
@timuric @peelar @karolkielecki

Hello everyone,

Our IT team decided to use Saleor as our e-commerce platform for our Marketplace project.

I have a question before we start developing it because our software will be more than a storefront.

**The Marketplace is one part of it and it's crucial to be very well integrated with Our Software**. We have a proprietary technology that we'll be reprogramming in Python. Which makes sense to use it together with Saleor.

Now, we must integrate the two software: **Saleor and Our Software**. They must work in sync because one will feed the other. And our users can't notice that they are moving from one to another. The surfing experience must be smooth. Meaning, they must log in on one page and have access to the marketplace (Saleor backend) and Our Software.

In other words, these two parts must be only one from the user's perspective/experience.

For the Frontend is easy to see this integration. But the Backend is not clear to me yet due to the specifics of Our Software.

Our Software will have a lot of specific features, tables, resources, images, links, pages, etc., that have nothing to do with the e-commerce/marketplace itself. At the same time, it will have a smooth connection with the marketplace. While surfing in Our Software, the user will be able to buy physical products, online courses, events, etc. that will be available in the marketplace at any time.

Perhaps, one of the features will be to host Online Courses, for example. So, the user will be able to watch the videos on Our Software they just bought from the marketplace that is running on Saleor.

It's not clear to me if the **Apps** extension would resolve all that. In short, our Software will need/have:

- Proprietary code in Python that has nothing to do with the marketplace.
- API for these specific methods that have nothing to do with the marketplace.
- The ability for the user to add a product to a cart or buy a product/go to checkout from "Our Software" using the same authenticated session.

My main question is: **What is the easiest and simplest way to achieve it?** _Note: considering that I don't want to touch Saleor's core to avoid future upgrades._

Other questions are:
1. How can I use Saleor as the "center of authentication/user data" and have the user logged in to "the Frontend and Backend" of Our Software as well?
2. Should I have another Backend application written in Python with our own API, or is there a way to extend Saleor on the Python and PostgreSQL level without causing problems for future upgrades?
3. Should I share the same database? Maybe using a different scheme not to mix things and to avoid issues with future Saleor upgrades.

I'm very excited to get started on this project, and I appreciate your constructive comments.

Thank you.

**Comment by peteror** [comment, marked as answer]
Hello, 

if your software does more than just e-commerce and is actually the center of the customer experience, here's what I would do: 

1. Build your frontend connect directly to your software and use your software to make calls to Saleor. This way you can use Saleor as the customer database / authentication engine and still have a consistent user experience. 
Here's what I would do:
your frontend -> your software -> Saleor
(I wouldn't put a lot of logic into a frontend) Consider Saleor another module / service of your software and communicate with it through Saleor APIs and webhooks. If you want, you can make calls from the frontend to Saleor directly if you need to. 
3. Don't customize Saleor, it will make upgrades difficult (I made that mistake)
4. This way you can write your software in whatever language (as I understand you already have it built in another language), as you'll connect to Saleor through HTTP API calls (which is not too bad for performance if the servers are close and you don't have very high - 100s / sec - call volumes)

**Reply by timuric** [reply]
@leandrobortolotto I fully agree with @peteror you should treat Saleor as a building block same as your CMS, payment gateway, or email provider

**Reply by leandrobortolotto** [reply]
@peteror @timuric Thank you very much for both of your replies. It makes sense!

Thank you.
